File: library/src/main/java/com/github/popalay/rxrealm/RxRealm.java
package com.github.popalay.rxrealm;

import android.os.HandlerThread;
import android.os.Process;

import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import io.realm.Realm;
import io.realm.RealmObject;
import io.realm.RealmResults;
import rx.Completable;
import rx.Observable;
import rx.Scheduler;
import rx.Single;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;
import rx.functions.Func1;

public class RxRealm {

    private RxRealm() {}

    public static <T extends RealmObject> Observable<List<T>> listenList
            (final Func1<Realm, RealmResults<T>> query) {
        final AtomicReference<Realm> realmReference = new AtomicReference<>(null);
        return listenRealmResults(query, realmReference)
                .map(result -> realmReference.get().copyFromRealm(result));
    }

    public static <T extends RealmObject> Observable<T> listenElement(final Func1<Realm, RealmResults<T>> query) {
        final AtomicReference<Realm> realmReference = new AtomicReference<>(null);
        return listenRealmResults(query, realmReference)
                .map(result -> !result.isEmpty() ? realmReference.get().copyFromRealm(result).get(0) : null);
    }

    public static <T extends RealmObject> Single<List<T>> getList(final Func1<Realm, RealmResults<T>> query) {
        return Single.fromEmitter(emitter -> {
            final Realm realm = Realm.getDefaultInstance();
            final RealmResults<T> realmResults = query.call(realm);
            emitter.onSuccess(realm.copyFromRealm(realmResults));
            emitter.setCancellation(realm::close);
        });
    }

    public static <T extends RealmObject> Single<T> getElement(final Func1<Realm, T> query) {
        return Single.fromEmitter(emitter -> {
            final Realm realm = Realm.getDefaultInstance();
            final T result = query.call(realm);
            if (result != null) {
                emitter.onSuccess(realm.copyFromRealm(result));
            } else {
                emitter.onSuccess(null);
            }
            emitter.setCancellation(realm::close);
        });
    }

    public static Completable doTransactional(final Action1<Realm> transaction) {
        return Completable.fromAction(() -> {
            try (Realm realm = Realm.getDefaultInstance()) {
                realm.executeTransaction(transaction::call);
            }
        });
    }

    private static <T extends RealmObject> Observable<RealmResults<T>> listenRealmResults(
            final Func1<Realm, RealmResults<T>> query, AtomicReference<Realm> realmReference) {
        final HandlerThread dbHandler = createDbHandler();
        final Scheduler scheduler = AndroidSchedulers.from(dbHandler.getLooper());
        return Observable.defer(() -> {
            final Realm realm = Realm.getDefaultInstance();
            realmReference.set(realm);
            return query.call(realm).asObservable();
        })
                .filter(result -> result.isLoaded() && result.isValid())
                .subscribeOn(scheduler)
                .unsubscribeOn(scheduler);
    }

    private static HandlerThread createDbHandler() {
        final HandlerThread handlerThread = new HandlerThread("RealmReadThread", Process.THREAD_PRIORITY_BACKGROUND);
        handlerThread.start();
        return handlerThread;
    }

}

File: library-kotlin/src/main/kotlin/com/popalay/rxrealm/RxRealm.kt
package com.popalay.rxrealm

import android.os.Handler
import android.os.HandlerThread
import android.os.Process
import io.realm.Realm
import io.realm.RealmObject
import io.realm.RealmResults
import rx.Observable
import rx.Single
import rx.android.schedulers.AndroidSchedulers
import java.util.concurrent.atomic.AtomicReference

class RxRealm private constructor() {

    fun <T : RealmObject> listenList(query: (Realm) -> RealmResults<T>): Observable<List<T>> {
        val dbHandler = createDbHandler()
        val scheduler = AndroidSchedulers.from(dbHandler.looper)
        val realmReference = AtomicReference<Realm>(null)
        return Observable.defer {
            val realm = Realm.getDefaultInstance()
            realmReference.set(realm)
            query.invoke(realm).asObservable()
        }
                .filter { result -> result.isLoaded && result.isValid }
                .map { result -> realmReference.get().copyFromRealm(result) }
                .subscribeOn(scheduler)
                .unsubscribeOn(scheduler)
                .doOnUnsubscribe { unsubscribeOnDbThread(dbHandler, realmReference.get()) }
    }

    fun <T : RealmObject> listenElement(query: (Realm) -> T): Observable<T> {
        val dbHandler = createDbHandler()
        val scheduler = AndroidSchedulers.from(dbHandler.looper)
        val realmReference = AtomicReference<Realm>(null)
        return Observable.defer {
            val realm = Realm.getDefaultInstance()
            realmReference.set(realm)
            query(realm).asObservable<T>()
        }
                .filter { result -> result.isLoaded && result.isValid }
                .map { result -> realmReference.get().copyFromRealm(result) }
                .subscribeOn(scheduler)
                .unsubscribeOn(scheduler)
                .doOnUnsubscribe { unsubscribeOnDbThread(dbHandler, realmReference.get()) }
    }

    fun <T : RealmObject> getList(query: (Realm) -> RealmResults<T>): Single<List<T>> {
        return Single.fromEmitter { emitter ->
            val realm = Realm.getDefaultInstance()
            emitter.onSuccess(realm.copyFromRealm(query.invoke(realm)))
            emitter.setCancellation { realm.close() }
        }
    }

    fun <T : RealmObject?> getElement(query: (Realm) -> T): Single<T> {
        return Single.fromEmitter { emitter ->
            val realm = Realm.getDefaultInstance()
            val result = query(realm)
            if (result != null) {
                emitter.onSuccess(realm.copyFromRealm(result))
            } else {
                emitter.onSuccess(null)
            }
            emitter.setCancellation { realm.close() }
        }
    }

    fun doTransactional(transaction: (Realm) -> Unit) {
        Realm.getDefaultInstance().use { it.executeTransaction { transaction(it) } }
    }

    private fun createDbHandler(): HandlerThread {
        val handlerThread = HandlerThread("RealmReadThread", Process.THREAD_PRIORITY_BACKGROUND)
        handlerThread.start()
        return handlerThread
    }

    private fun unsubscribeOnDbThread(handlerThread: HandlerThread, realm: Realm) {
        val handler = Handler(handlerThread.looper)
        handler.post {
            realm.close()
            handlerThread.quitSafely()
        }
    }
}